### PR TITLE
[CDAP-4242] update mapreduce and other program sections about getting datasets

### PR DESCRIPTION
**Pull request by anew**
These docs were still explaining static datasets, but now we do not require @UseDataset anymore. Also restructures the docs a little bit: moved the "Datasets and MapReduce" into the MapReduce section. It makes more sense there, it seemed out of place in the Datasets section, especially because MapReduce had not been introduced at that time.

### CODE REVIEW

**Comment by anew** [comment]
This only changes docs. Quick build succeeded here: http://builds.cask.co/browse/CDAP-DOB43-7, and the generated HTML is here: http://builds.cask.co/artifact/CDAP-DOB43/shared/build-7/Docs-HTML/3.3.0-SNAPSHOT/en/index.html

**Comment by awholegunch** [comment]
I've re-run the [Quick Build](http://builds.cask.co/browse/CDAP-DOB43-8). Failed because of a change of the cdap-default.xml. However, the build publishes anyways.

Relevant pages:

- [Datasets](http://builds.cask.co/artifact/CDAP-DOB43/shared/build-8/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/building-blocks/datasets/index.html)
- [System and Custom Datasets](http://builds.cask.co/artifact/CDAP-DOB43/shared/build-8/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/building-blocks/datasets/system-custom.html)
- [MapReduce Programs](http://builds.cask.co/artifact/CDAP-DOB43/shared/build-8/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/building-blocks/mapreduce-programs.html)
- [Services](http://builds.cask.co/artifact/CDAP-DOB43/shared/build-8/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/building-blocks/services.html)
- [Spark](http://builds.cask.co/artifact/CDAP-DOB43/shared/build-8/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/building-blocks/spark-programs.html)

**Comment by awholegunch** [comment]
I've made some additional changes to support this. Please review my commits. Otherwise, LGTM.

File: cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst [hunk 1]
```diff
@@ -161,12 +182,147 @@ declaration and (2) an injection:
...
+
+  public class KeyValueTable extends AbstractDataset
+                             implements BatchReadable<byte[], byte[]> {
+    ...
+    public List<Split> getSplits(int numSplits, byte[] start, byte[] stop);
+  }
+
+To read a range of keys and give a hint that you want 16 splits, write::
+
+  @Override
```
Comment (anwar6953) [inline cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: I think this is a concern with the existing documentation, but I think the `@Override` should be moved down to right above the `public void beforeSubmit(MapReduceContext context) throws Exception {`.

Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: Probably a copy/paste lapse. Good catch!

**Comment by anew** [comment]
@awholegunch your changes look good to me. I will merge this after fixing the issue that Ali found.

**Pull request merged**